Update reply block rendering based on position

obenland opened the pull request:
Fixes #2124

## Proposed changes:
* Transform reply blocks to @-mention links only when they are the first block in a post for ActivityPub content
* Refactor block rendering logic from Post transformer to Blocks class for better organization
* Add dynamic callback registration during ActivityPub content generation
* Deprecate old methods in Post transformer while maintaining backward compatibility

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Create a new WordPress post with a reply block as the first block
* Publish the post and view it via ActivityPub (e.g., using the ActivityPub REST API endpoint)
* Verify the reply block is rendered as an @-mention link in ActivityPub content
* Create another post with a reply block that is NOT the first block
* Verify this reply block is rendered as a regular clickable link in ActivityPub content
* Test that local rendering remains consistent (shows clickable links in both cases)

### Changelog entry

- [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [x] Patch

#### Type

- [x] Fixed - for any bug fixes

#### Message

Fix reply block rendering inconsistency where blocks were always converted to @-mentions in ActivityPub content. Now only first reply blocks become @-mentions, others remain as regular links.

</details>

obenland committed:
Fix reply block rendering inconsistency based on position

Transform reply blocks to @-mention links only when they are the first block in a post for ActivityPub content. This provides consistent behavior where:
- First reply block = @-mention (indicating the entire post is a reply)
- Non-first reply blocks = regular links (referencing other posts)

Refactor block rendering logic from Post transformer to Blocks class and add dynamic callback registration during ActivityPub content generation.

Fixes #2124

matticbot committed:
Add changelog

obenland committed:
Remove unused Post import in test blocks

Deleted the unused import of Activitypub\Transformer\Post from class-test-blocks.php to clean up the code.

obenland committed:
Merge branch 'update/reply-link' of https://github.com/Automattic/wordpress-activitypub into update/reply-link

pfefferle commented:
Can you please describe the problem a bit and how this PR fixes it?

obenland commented:
#2124 pointed out how remote representations of the federated reply block always get rendered as a mention, regardless of whether the block is the first block in a post or somewhere in the middle of post. This can lead to awkward renderings:

<img width="348" height="354" alt="Image" src="https://github.com/user-attachments/assets/0e74babf-7be9-412b-abf2-da06a83d8272" />

This PR [adds a check](https://github.com/Automattic/wordpress-activitypub/pull/2132/files#diff-89d47d0bcb43bf40e33a19f8e947c9942d5f8cc07882ea949a55eeecff35a2b7R396) in a callback that's hooked into the post transformer to only convert the reply block display to a mention-type handle when it's the first block in a post.
If that is not the case, [it will render the URL](https://github.com/Automattic/wordpress-activitypub/pull/2132/files#diff-89d47d0bcb43bf40e33a19f8e947c9942d5f8cc07882ea949a55eeecff35a2b7R236).

pfefferle committed:
Merge branch 'trunk' into update/reply-link

pfefferle commented:
The cases you described, work as expected, but I thought I save some time and test all cases in one post :)

I do not know if that is a usecase, but if you have two reply blocks in one post, one as the first block and one later in the post, both will still show up as @-replies.

<img width="1272" height="692" alt="Screenshot 2025-09-10 at 10 52 22" src="https://github.com/user-attachments/assets/986377bb-2cdb-48a0-961e-3a80077f3c52" />

obenland committed:
Merge branch 'trunk' into update/reply-link

obenland committed:
Add failing test for multiple reply blocks behavior

Test demonstrates that when multiple reply blocks exist in a post, currently both get transformed to @-mentions instead of only the first one. This test will pass once the fix is implemented to ensure only the first reply block becomes an @-mention while subsequent ones remain as regular links.

obenland committed:
Fix multiple reply blocks transformation issue

Only transform the first reply block to @-mention by removing the filter hook after first execution. This ensures subsequent reply blocks in the same post remain as regular links instead of being incorrectly transformed to @-mentions.

The fix uses a self-removing hook approach where generate_reply_link() unhooks itself on first execution, preventing transformation of additional reply blocks.

obenland commented:
@pfefferle This is ready for another round

pfefferle reviewed (approved):
oops, forget that one!

works like a charm!

pfefferle committed:
Merge branch 'trunk' into update/reply-link
